text/0000-canonical-metadata.md

```diff
@@ -0,0 +1,1466 @@
+    location: SourceLocation;
+    hints?: DynamicImportHint[];
+}
+interface DynamicImportHint {
+    // Hint value with leading and trailing spaces of hint phrase trimmed
+    rawValue: string;
+    key: string;
+    value: string;
+    location: SourceLocation;
+}
+
+// Information about a export statement in a module.
+interface Export {
+    exportsList?: {
+        type: 'class' | 'function' | 'expression';
+        name?: 'string';
+    }[];
+    default?: boolean;
+    location: SourceLocation;
+}
+
+// Information about an export statement used to reexport bindings imported from another module.
+interface ReExport {
+    exportsList: {
+        name: string;
+        type: 'class' | 'function' | 'expression';
+    }[];
+    // module name in canonical form
+    moduleSpecifier: string;
+    location: SourceLocation;
+}
+
+// Information about a dom event instantiated in script.
+interface DOMEvent {
+    eventType: string;
+    isCustomEvent?: boolean;
+    options?: {
+        bubbles?: boolean;
+        composed?: boolean;
+    };
+    location: SourceLocation;
+}
+
+// Information about an event handled declaratively in the template.
+interface ProgrammaticEventListener {
+    type: string;
+    targetType: 'host' | 'shadowRoot' | 'Node';
+    options?: {
+        capture?: boolean;
+    };
+    location: SourceLocation;
+}
+
+// A parent component class including the 'LightningElement' class provided by the lwc module.
+type ParentClass =
+    | string // Another class declared in the same file
+    | Mixin
+    // or a class imported from an external module
+    | {
+          name: string;
+          // module name in full canonical form, the ModuleReference object can be looked up at 
+          // the root ScriptFile object using this name
+          moduleName?: string;
+          // Location of the parent class identifier
+          location: SourceLocation;
+      };
+
+// Information about a module imported into a module using the 'import' statement.
+interface ModuleReference {
+    // Full canonical name, eg. lightning/input, lightning/recordForm
+    name: string;
+    namespace?: string;
+    moduleIdentifier?:
+        | 'apexClass'
+        | 'apexMethod'
+        | 'apexContinuation'
+        | 'client'
+        | 'community'
+        | 'component'
+        | 'contentAssetUrl'
+        | 'customPermission'
+        | 'dynamicComponent'
+        | 'slds'
+        | 'messageChannel'
+        | 'i18n'
+        | 'gate'
+        | 'label'
+        | 'metric'
+        | 'module'
+        | 'internal'
+        | 'resourceUrl'
+        | 'schema'
+        | 'sobjectClass'
+        | 'sobjectField'
+        | 'user'
+        | 'userPermission';
+    // whether the module reference is the built in lwc module or the @salesforce scoped module or
+    // a local file in the current bundle or an module imported from an external source(eg. a component)
+    type: 'lwc' | '@salesforce' | 'internal' | 'external';
```

Comment:
what are the last two types for? Today, the notion of 'internal' vs 'platform' is specific to compilation and are tied to the bundle not a reference.